### PR TITLE
[Dynamo] Match closures by code ID

### LINKED ISSUE #107866: [Dynamo] Avoid id test on closures
### 🐛 Describe the bug

Suppose you have:

```
def f(x):
  def g(y):
    ...
  graph_break()
  ... some stuff including g ...
```

The continuation frame after the graph break will always recompile, because g is a closure and reallocated every call to f, but we will typically install an ID_MATCH on it.

It would be better if we could introspect into closures and explicitly guard on the freevars bound on it, which may turn out to be constant from run to run.

This showed up in a Meta internal model. An alternative fix is to diagnose why the graph break was happening and fix that instead.

### Versions

main

cc @msaroufim @wconstab @bdhirsh @anijain2305 @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov

### PR DESCRIPTION
**Pull request by Fidget-Spinner**
Closes https://github.com/pytorch/pytorch/issues/107866

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @chenyang78 @aakhundov @kadeng

### CODE REVIEW

**Review by ezyang — CHANGES_REQUESTED** [review]
Don't you also have to guard on the free values of the closure too?

**Comment by Fidget-Spinner** [comment]
> Don't you also have to guard on the free values of the closure too?

I was under the impression that dynamo inlines the closure. So its free values are guarded by the guards of the caller. That assumption might be wrong though.

**Comment by Fidget-Spinner** [comment]
The failing tests are due to an unrelated bug, which I fixed in https://github.com/pytorch/pytorch/issues/109458. The only reason why it's failing now is that this PR allows dynamo to execute more code, so more bugs get uncovered.

**Comment by ezyang** [comment]
This might be OK but I am a little nervous because the guards are associated with the function objects at construction time, and you could imagine that we might not inline into the closure (in which case I think your guard is correct) but do something  different that wouldn't trigger inlining (e.g., pulling out `__self__` from a bound method). I suspect as Dynamo coverage is today this cannot be triggered. Added some other reviewers for a second opinion.

**Review by jansel — CHANGES_REQUESTED** [review]
I think we will need to guard on the cell vars inside the closure.

We may already guard on these due to inlining, but we need to copy these guards to the function object since the function might get passed to something else rather than inlined.

**Comment by Fidget-Spinner** [comment]
Ok, I played around with this for a few more hours (including writing code to guard the free variables), I have concluded we indeed don't need the free variable check. A scratch proof by exhaustion:

1. Case 1: the closure is inlined, it is treated as a part of the code object and thus has guards protecting it.
2. Case 2: the closure is not inlined, due to a graph break or whatever. **Dynamo will trace into the closure and treat it as its own code object, which comes with guards for its own values**. Thus it is still safe.

I managed to show a case of 2. occuring with the following code snippet:

```py
import torch

@torch.compile()
def program(x, y):
  one = lambda x, y: x + y
  def inner():
    # Force no inlining
    torch._dynamo.graph_break()
    return one(x, y)
  res = inner()
  one = lambda x, y: x - y
  res += inner()
  return res

input1 = torch.randn(1)
input2 = torch.randn(1)

print(program(input1, input2), input1 + input1)
```

It gives the following guards+output:

```
GUARDS
 if not (___guarded_code.valid):
    return False
if not (___check_global_state()):
    return False
_var0 = L['x']
if not (hasattr(_var0, '_dynamo_dynamic_indices') == False):
    return False
_var1 = L['y']
if not (hasattr(_var1, '_dynamo_dynamic_indices') == False):
    return False
if not (utils_device.CURRENT_DEVICE == None):
    return False
if not (___skip_backend_check() or ___current_backend() == ___lookup_backend(139671583626576)):
    return False
if not (___check_tensors(_var0, _var1, tensor_check_names=tensor_check_names)):
    return False

GUARDS
 if not (___guarded_code.valid):
    return False
if not (___check_global_state()):
    return False
_var0 = L['x']
if not (hasattr(_var0, '_dynamo_dynamic_indices') == False):
    return False
_var1 = L['y']
if not (hasattr(_var1, '_dynamo_dynamic_indices') == False):
    return False
if not (utils_device.CURRENT_DEVICE == None):
    return False
if not (___skip_backend_check() or ___current_backend() == ___lookup_backend(139671583626576)):
    return False
if not (___check_tensors(_var0, _var1, tensor_check_names=tensor_check_names)):
    return False

GUARDS
 if not (___guarded_code.valid):
    return False
if not (___check_global_state()):
    return False
_var0 = L['x']
if not (hasattr(_var0, '_dynamo_dynamic_indices') == False):
    return False
_var1 = L['y']
if not (hasattr(_var1, '_dynamo_dynamic_indices') == False):
    return False
_var2 = L['one']
if not (hasattr(_var2, '__code__')):
    return False
if not (___check_obj_id(_var2.__code__, 139671583431424)):
    return False
if not (utils_device.CURRENT_DEVICE == None):
    return False
if not (___skip_backend_check() or ___current_backend() == ___lookup_backend(139671583626576)):
    return False
if not (___check_tensors(_var0, _var1, tensor_check_names=tensor_check_names)):
    return False

GUARDS
 if not (___guarded_code.valid):
    return False
if not (___check_global_state()):
    return False
_var0 = L['x']
if not (hasattr(_var0, '_dynamo_dynamic_indices') == False):
    return False
_var1 = L['y']
if not (hasattr(_var1, '_dynamo_dynamic_indices') == False):
    return False
_var2 = L['inner']
if not (hasattr(_var2, '__code__')):
    return False
if not (___check_obj_id(_var2.__code__, 139671584764448)):
    return False
_var3 = L['___stack0']
if not (hasattr(_var3, '_dynamo_dynamic_indices') == False):
    return False
if not (utils_device.CURRENT_DEVICE == None):
    return False
if not (___skip_backend_check() or ___current_backend() == ___lookup_backend(139671583626576)):
    return False
if not (___check_tensors(_var0, _var1, _var3, tensor_check_names=tensor_check_names)):
    return False

GUARDS
 if not (___guarded_code.valid):
    return False
if not (___check_global_state()):
    return False
_var0 = L['x']
if not (hasattr(_var0, '_dynamo_dynamic_indices') == False):
    return False
_var1 = L['y']
if not (hasattr(_var1, '_dynamo_dynamic_indices') == False):
    return False
_var2 = L['one']
if not (hasattr(_var2, '__code__')):
    return False
if not (___check_obj_id(_var2.__code__, 139671583432464)):
    return False
if not (utils_device.CURRENT_DEVICE == None):
    return False
if not (___skip_backend_check() or ___current_backend() == ___lookup_backend(139671583626576)):
    return False
if not (___check_tensors(_var0, _var1, tensor_check_names=tensor_check_names)):
    return False

GUARDS
 if not (___guarded_code.valid):
    return False
if not (___check_global_state()):
    return False
_var0 = L['x']
if not (hasattr(_var0, '_dynamo_dynamic_indices') == False):
    return False
_var1 = L['y']
if not (hasattr(_var1, '_dynamo_dynamic_indices') == False):
    return False
_var2 = L['___stack0']
if not (hasattr(_var2, '_dynamo_dynamic_indices') == False):
    return False
_var3 = L['___stack1']
if not (hasattr(_var3, '_dynamo_dynamic_indices') == False):
    return False
if not (utils_device.CURRENT_DEVICE == None):
    return False
if not (___skip_backend_check() or ___current_backend() == ___lookup_backend(139671583626576)):
    return False
if not (___check_tensors(_var0, _var1, _var2, _var3, tensor_check_names=tensor_check_names)):
    return False

tensor([0.4222]) tensor([0.4222])
```

File: torch/_dynamo/guards.py
Comment (jansel) [inline torch/_dynamo/guards.py:499]:
```suggestion
                    f"___check_obj_id(getattr({ref}, '__code__', None), {self.id_ref(val.__code__)})",
```

**Review by jansel — CHANGES_REQUESTED** [review]

**Review by ezyang — APPROVED** [review]
Thanks a lot!

**Review by jansel — APPROVED** [review]

File: torch/_dynamo/guards.py
Comment (voznesenskym) [inline torch/_dynamo/guards.py:490]: I don't get how this is sound, code id is not enough?

**Comment by voznesenskym** [comment]
@Fidget-Spinner I'm going to dig in here and see if we can recover this PR with a few extra safety checks.

File: torch/_dynamo/guards.py
Comment (Fidget-Spinner) [reply torch/_dynamo/guards.py:490]: It's safe because of the reasoning in here https://github.com/pytorch/pytorch/pull/109427#issuecomment-1732538371

**Comment by Fidget-Spinner** [comment]
@voznesenskym the CI fails because there's a bug in how dynamo represents lambdas (it just compiles it down to `torch.cuda.<lambda>`). This code path likely wasn't triggered before because this PR enables more lambdas to run. The failure is unrelated to safety.

File: torch/_dynamo/guards.py
Comment (Fidget-Spinner) [reply torch/_dynamo/guards.py]: I found that we can't use `getattr` here because `getattr` internally raises an error in CPython and suppresses it, which leads the guard function to think it has errored wrongly.

**Comment by Fidget-Spinner** [comment]
Ok after fixing the compile bug, there's now a weird bug where the symbolic evaluator thinks a `functools.partial` is a tensor and tries to do a symbolic check on it (which raises an error).

**Comment by Fidget-Spinner** [comment]
Ok I fixed the bug. The codegen was missing some guards for `functools.partial`. It was some guard set mutability issue. Every time you mutate a `VariableTracker` you need to return a new copy as it's supposed to be immutable. However, `FunctoolsPartialVariable` was subclassing `VariableTracker` and then updating the guards set after initializing it. This led to some wonky stuff happening that I can't explain but have fixed by ensuring immutability.

**Comment by Fidget-Spinner** [comment]
I still can't fix the bug. For some reason, some of the guards checking for key equality of `functools.partial` are dropped. Leading to the guard to access non-existent keys in a dictionary.

**Comment by ezyang** [comment]
@Fidget-Spinner do you know if the guard is non-existent, or if it's being ordered incorrectly? @jansel told me the way that we make sure guards run in the right order today is we sort them by length, so the keys test should have shorter length than the inner dim test. But this is pretty fragile and it wouldn't surprise me if we messed it up.

Relevant code:

```
    def sort_key(self):
        return (
            self.source.value if self.source else -1,
            len(self.name),
            self.name,
            self.inner_create_fn().__code__.co_firstlineno,
        )
```

**Comment by Fidget-Spinner** [comment]
> @Fidget-Spinner do you know if the guard is non-existent, or if it's being ordered incorrectly?

It's present for some cases, but disappears altogether specifically after a graph break, and only on a SymNodeVariable not ConstantVariable. I saw that the guards are using sets (unordered) as well, and went to check that but that isn't the cause.

I had a long pair session with @ysiraichi yesterday, and we still couldn't figure out why/create a minimal reproducer.

**Comment by voznesenskym** [comment]
> I finally have a minimal reproducer, thanks to help from Yukio:
> 
> ```python
> import torch
> from functools import partial
> 
> 
> def bar(x, **kwargs):
>    return x + x
> 
> @torch.compile(dynamic=True)
> def foo(x, i):
>     def inner():
>         print("this is a graph_break")
>         return op(x)
>     op = partial(bar, dim=10)
>     x = inner()
>     op = partial(bar, other=10)   
>     return inner() + x
> 
> foo(torch.rand(1), 10)
> ```

Perfect. Looking.

**Comment by Fidget-Spinner** [comment]
Now that we know the bug is in `functools.partial` and not this PR, I will unblock the PR by removing the buggy use in that specific test. I have created another issue to track the bug https://github.com/pytorch/pytorch/issues/110790.

**Comment by Fidget-Spinner** [comment]
I'm on time off this week, but plan to merge this once I'm back. @jansel and @ezyang please let me know if you're happy with the PR. I have fixed all the breaking tests (or at least, created issues for them), so this won't kill the CI again in the future (hopefully).

File: torch/fx/node.py
Comment (ezyang) [inline torch/fx/node.py:96]: @Chillee could you look at this?

Comment (ezyang) [reply torch/fx/node.py:96]: In particular, I'm wondering if the name we generate here must actually be "dereferenceable" (e.g., eval(_get_qualified_name(func)) is func`

**Review by ezyang — APPROVED** [review]
I am not sure how likely the functools.partial bug is likely to hit in practice. It may be safer to fix that bug first before landing this. Giving the re-stamp for now though.

**Comment by Fidget-Spinner** [comment]
The functools partial bug has been resolved upstream. I will be merging this after CI passes.

**Pull request closed without merging**
